### PR TITLE
 Gauge: Add support for constant labels

### PR DESCRIPTION
**Pull request by mayurkale22**
This is as per the specs: https://github.com/census-instrumentation/opencensus-specs/pull/248

### CODE REVIEW

File: packages/opencensus-core/src/common/validations.ts [hunk 1]
```diff
@@ -43,3 +45,27 @@ export function validateArrayElementsNotNull<T>(
     throw new Error(`${errorMessage} elements should not be a NULL`);
   }
 }
+
+/** Throws an error if any of the map elements is null. */
+export function validateMapElementNotNull<T>(
+    map: Map<T, T>, errorMessage: string) {
+  for (const [key, value] of map.entries()) {
+    if (key === null || key === undefined || value === null ||
```
Comment (draffensperger) [inline packages/opencensus-core/src/common/validations.ts]: Optional: you can use `== null` to test for both null and undefined, which is a permissible exception IMO to general (good) advice to always use `===`. See https://stackoverflow.com/questions/2559318/how-to-check-for-an-undefined-or-null-variable-in-javascript for details.

File: packages/opencensus-core/src/common/validations.ts [hunk 2]
```diff
@@ -43,3 +45,27 @@ export function validateArrayElementsNotNull<T>(
...
+        value === undefined) {
+      throw new Error(`${errorMessage} elements should not be a NULL`);
+    }
+  }
+}
+
+/** Throws an error if any of the array element present in the map. */
+export function validateDuplicateKeys(
+    keys: LabelKey[], map: Map<LabelKey, LabelValue>) {
+  const keyNames: Set<string> = new Set();
```
Comment (draffensperger) [inline packages/opencensus-core/src/common/validations.ts]: Optional: could this whole block become just be something like `const keysAndConstantKeys = new Set([...keyNames, ...map.keys].map(k => k.key))`?

File: packages/opencensus-core/src/common/validations.ts [hunk 3]
```diff
@@ -43,3 +45,27 @@ export function validateArrayElementsNotNull<T>(
...
+    if (key === null || key === undefined || value === null ||
+        value === undefined) {
+      throw new Error(`${errorMessage} elements should not be a NULL`);
+    }
+  }
+}
+
+/** Throws an error if any of the array element present in the map. */
+export function validateDuplicateKeys(
+    keys: LabelKey[], map: Map<LabelKey, LabelValue>) {
```
Comment (draffensperger) [inline packages/opencensus-core/src/common/validations.ts]: Should `map` be called `constantLabels` here?

File: packages/opencensus-core/src/metrics/gauges/derived-gauge.ts [hunk 4]
```diff
@@ -94,12 +95,19 @@ export class DerivedGauge implements types.Meter {
...
+   *     labels for the Metric.
    */
   constructor(
       name: string, description: string, unit: string,
-      type: MetricDescriptorType, labelKeys: LabelKey[]) {
-    this.metricDescriptor = {name, description, unit, type, labelKeys};
+      type: MetricDescriptorType, labelKeys: LabelKey[],
+      readonly constantLabels: Map<LabelKey, LabelValue>) {
     this.labelKeysLength = labelKeys.length;
+    const combinedLabelKeys = labelKeys.concat(...constantLabels.keys());
```
Comment (draffensperger) [inline packages/opencensus-core/src/metrics/gauges/derived-gauge.ts]: I notice below you use `[...gaugeEntry.labelValues, ...this.constantLabelValues]`, so should this be `[...labelKeys, ...consantLabels.keys()]` to match?

File: packages/opencensus-core/src/common/validations.ts [hunk 5]
```diff
@@ -43,3 +45,27 @@ export function validateArrayElementsNotNull<T>(
...
+  const keyNames: Set<string> = new Set();
+  keys.forEach(key => keyNames.add(key.key));
```
Comment (songy23) [inline packages/opencensus-core/src/common/validations.ts]: nit: consider adding a separate check on duplicated keys in `keys`:
```
if (keyNames.size !== keys.length) {
  throw new Error(...);
}
```

File: packages/opencensus-core/src/metrics/gauges/derived-gauge.ts (hunk 4 above)
Comment (mayurkale22) [reply packages/opencensus-core/src/metrics/gauges/derived-gauge.ts]: Good Idea, Done.

File: packages/opencensus-core/src/common/validations.ts (hunk 3 above)
Comment (mayurkale22) [reply packages/opencensus-core/src/common/validations.ts]: Sure

File: packages/opencensus-core/src/common/validations.ts (hunk 2 above)
Comment (mayurkale22) [reply packages/opencensus-core/src/common/validations.ts]: Done

File: packages/opencensus-core/src/common/validations.ts (hunk 5 above)
Comment (mayurkale22) [reply packages/opencensus-core/src/common/validations.ts]: For simplicity, I have combined with existing check.

File: packages/opencensus-core/src/common/validations.ts (hunk 1 above)
Comment (mayurkale22) [reply packages/opencensus-core/src/common/validations.ts]: SGTM, done. Thanks for sharing the link.

**Pull request merged**